controllers/etcd_controller.go

```diff
@@ -576,64 +613,93 @@ func (r *EtcdReconciler) reconcileStatefulSet(cm *corev1.ConfigMap, svc *corev1.
 		for i := 1; i < len(filteredStatefulSets); i++ {
 			ss := filteredStatefulSets[i]
 			if err := r.Delete(context.TODO(), ss); err != nil {
-				logger.Error(err, "Error in deleting duplicate StatefulSet")
+				logger.Error(err, "error in deleting duplicate statefulset")
 				continue
 			}
 		}
 
 		// Return the updated statefulset
 		ss := &appsv1.StatefulSet{}
 		if err := r.Get(context.TODO(), types.NamespacedName{Name: filteredStatefulSets[0].Name, Namespace: filteredStatefulSets[0].Namespace}, ss); err != nil {
-			return nil, err
+			return nil, &reconcileResultRequeue, err
+		}
+
+		pvcList := &corev1.PersistentVolumeClaimList{}
+		if handleScaleDown {
+			ctx, cancel := context.WithTimeout(context.TODO(), DefaultTimeout)
+			defer cancel()
+			pvcList, err = r.prepareEtcdForScaleDown(ctx, etcd, ss, svc)
+			if err != nil {
+				return nil, &reconcileResultRequeueAfter, err
+			}
 		}
 
 		// Statefulset is claimed by for this etcd. Just sync the specs
 		if ss, err = r.syncStatefulSetSpec(ss, cm, svc, etcd, values); err != nil {
-			return nil, err
+			return nil, &reconcileResultRequeue, err
 		}
 
 		// restart etcd pods in crashloop backoff
 		selector, err := metav1.LabelSelectorAsSelector(ss.Spec.Selector)
 		if err != nil {
 			logger.Error(err, "error converting statefulset selector to selector")
-			return nil, err
+			return nil, &reconcileResultRequeue, err
 		}
 		podList := &v1.PodList{}
 		if err := r.List(context.TODO(), podList, client.InNamespace(etcd.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
-			return nil, err
+			return nil, &reconcileResultRequeue, err
 		}
 
 		for _, pod := range podList.Items {
 			if utils.IsPodInCrashloopBackoff(pod.Status) {
 				if err := r.Delete(context.TODO(), &pod); err != nil {
```

Comment:
It would be better to aggregate the errors in the loop and then `reconcileResultRequeueAfter` once after the loop. This way when multi-node supports gets in we can get deletes for all crashing pods in one reconciliation instead of potentially one deletion per reconciliation in the worst case.